community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/plans/rewriter/removeIdenticalPlans.scala

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.planner.logical.plans.rewriter
+
+import org.neo4j.cypher.internal.compiler.v2_3._
+import org.neo4j.cypher.internal.compiler.v2_3.ast.Expression
+import org.neo4j.cypher.internal.compiler.v2_3.planner.logical.plans._
+import org.neo4j.cypher.internal.compiler.v2_3.planner.{CardinalityEstimation, QueryGraph}
+
+/**
+ * Runs through LogicalPlan and calls copies each plan to make sure the
+ * plan doesn't contain elements that are referentially identical.
+ */
+case object removeIdenticalPlans extends Rewriter {
+
+  def apply(input: AnyRef) = bottomUp(instance).apply(input)
+
+  private val instance: Rewriter = Rewriter.lift {
+    case plan: LogicalPlan => plan.copyPlan()
```

Comment:
Once this has been applied to the bottom level, applying it to higher levels will be superfluous, right? because the `BottomUpRewriter` will take care of that copying... So applying this `copyPlan` will in fact create yet another copy... Seems like excessive copying to me...